test/node/object_id_tests.js

```diff
@@ -24,6 +25,195 @@ describe('ObjectId', function () {
+    var objectValidStringId1 = {
+      id: 'aaaaaaaaaaaaaaaaaaaaaaaa'
+    };
+    var objectValidStringId2 = {
+      id: 'abcdefghijkl'
+    };
+    var buf1 = Buffer.from('aaaaaaaaaaaaaaaaaaaaaaaa', 'hex');
+    var buf2 = Buffer.from('abcdefghijkl', 'hex');
+    expect(() => new ObjectId(objectValidStringId1).toHexString()).to.not.throw(TypeError);
+    expect(Buffer.from(new ObjectId(objectValidStringId1).id).equals(buf1));
+    expect(() => new ObjectId(objectValidStringId2).toHexString()).to.not.throw(TypeError);
+    expect(Buffer.from(new ObjectId(objectValidStringId2).id).equals(buf2));
+  });
+
+  it('should correctly create ObjectId from object with valid string id and toHexString method', function () {
+    function newToHexString() {
+      return 'BBBBBBBBBBBBBBBBBBBBBBBB';
+    }
+    var buf = Buffer.from('BBBBBBBBBBBBBBBBBBBBBBBB', 'hex');
+    var objectValidStringId1 = {
+      id: 'aaaaaaaaaaaaaaaaaaaaaaaa',
+      toHexString: newToHexString
+    };
+    var objectValidStringId2 = {
+      id: 'abcdefghijkl',
+      toHexString: newToHexString
+    };
+    expect(() => new ObjectId(objectValidStringId1).toHexString()).to.not.throw(TypeError);
+    expect(Buffer.from(new ObjectId(objectValidStringId1).id).equals(buf));
+    expect(() => new ObjectId(objectValidStringId2).toHexString()).to.not.throw(TypeError);
+    expect(Buffer.from(new ObjectId(objectValidStringId2).id).equals(buf));
+  });
+
+  it('should correctly create ObjectId from object with valid Buffer id', function () {
+    var validBuffer1 = Buffer.from('AAAAAAAAAAAAAAAAAAAAAAAA', 'hex');
+    var validBuffer2 = Buffer.from([1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12]);
+    var objectBufferId = {
+      id: validBuffer1
+    };
+    var objectBufferFromArray = {
+      id: validBuffer2
+    };
+    expect(() => new ObjectId(objectBufferId).toHexString()).to.not.throw(TypeError);
+    expect(Buffer.from(new ObjectId(objectBufferId).id).equals(validBuffer1));
+    expect(() => new ObjectId(objectBufferFromArray).toHexString()).to.not.throw(TypeError);
+    expect(Buffer.from(new ObjectId(objectBufferId).id).equals(validBuffer2));
+  });
+
+  it('should correctly create ObjectId from object with valid Buffer id and toHexString method', function () {
+    var validBuffer1 = Buffer.from('AAAAAAAAAAAAAAAAAAAAAAAA', 'hex');
+    var validBuffer2 = Buffer.from([1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12]);
+    var bufferToHex = Buffer.from('BBBBBBBBBBBBBBBBBBBBBBBB', 'hex');
+    function newToHexString() {
+      return 'BBBBBBBBBBBBBBBBBBBBBBBB';
+    }
+    var objectBufferId = {
+      id: validBuffer1,
+      toHexString: newToHexString
+    };
+    var objectBufferFromArray = {
+      id: validBuffer2,
+      toHexString: newToHexString
+    };
+    expect(() => new ObjectId(objectBufferId).toHexString()).to.not.throw(TypeError);
+    expect(Buffer.from(new ObjectId(objectBufferId).id).equals(bufferToHex));
```

Comment:
Maybe this isn't the intention of the test, but I remember we wanted to have a test show that the toHexString takes precedent. I think here we have an objectId made up of 'AAAA' but the toHexString returns 'BBB' so we can add a check that the resulting ObjectId _**is not**_ made up of 'A's

Reply:
Resolved offline after discussion